flatgfa/src/packedseq.rs

```diff
@@ -45,60 +58,88 @@ impl From<Nucleotide> for char {
 /// A compressed vector-like structure for storing nucleotide sequences
 ///     - Two base pairs are stored per byte
 ///
-pub struct PackedVec {
-    /// A vector that stores a compressed encoding of this PackedVec's sequence
+pub struct PackedSeqStore {
+    /// A vector that stores a compressed encoding of this PackedSeqStore's sequence
     data: Vec<u8>,
 
     /// True if the final base pair in the sequence is stored at a
     ///                   high nibble
     high_nibble_end: bool,
 }
 
-impl PackedVec {
-    /// Creates a new empty PackedVec
-    pub fn new() -> Self {
-        PackedVec {
-            data: Vec::new(),
-            high_nibble_end: true,
-        }
+pub struct PackedSeqView<'a> {
+    data: &'a [u8],
+
+    /// True if the final base pair in the sequence is stored at a
+    ///                   high nibble
+    high_nibble_end: bool,
+}
+
+#[derive(FromBytes, FromZeroes, AsBytes, Debug)]
+#[repr(packed)]
+pub struct PackedToc {
+    magic: u64,
+    data: Size,
+    high_nibble_end: u8,
+}
+
+impl PackedToc {
+    pub fn size(&self) -> usize {
+        size_of::<Self>() + self.data.bytes::<u8>()
     }
 
-    /// Returns a compressed PackedVec given an uncompressed vector `arr`
-    pub fn create(arr: Vec<Nucleotide>) -> Self {
-        let mut new_vec = PackedVec::new();
-        for item in arr {
-            new_vec.push(item);
+    fn full(seq: &PackedSeqView) -> Self {
+        Self {
+            magic: MAGIC_NUMBER,
+            data: Size {
+                len: seq.data.len(),
+                capacity: seq.data.len(),
+            },
+            high_nibble_end: if seq.high_nibble_end { 1u8 } else { 0u8 },
         }
-        new_vec
     }
+}
 
-    /// Appends `input` to the end of this PackedVec
-    pub fn push(&mut self, input: Nucleotide) {
-        let value = input.into();
-        assert!(value <= 0xF);
-        if self.high_nibble_end {
-            self.data.push(value);
-            self.high_nibble_end = false;
-        } else {
-            let last_index = self.data.len() - 1;
-            self.data[last_index] |= value << 4;
-            self.high_nibble_end = true;
+fn read_packed_toc(data: &[u8]) -> (&PackedToc, &[u8]) {
```

Comment:
It seems like this should maybe be a method on `PackedToc`, with `Self` in the return type. It could then have a simpler name, like `read`.